src/go/log/log_test.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2021 The Magma Authors.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLevel_String(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		level Level
+		want  string
+	}{
+		{
+			level: DebugLevel,
+			want:  "DEBUG",
+		},
+		{
+			level: InfoLevel,
+			want:  "INFO",
+		},
+		{
+			level: WarnLevel,
+			want:  "WARN",
+		},
+		{
+			level: ErrorLevel,
+			want:  "ERROR",
+		},
+		{
+			level: 100,
+			want:  "INVALID (100)",
+		},
+	}
+
+	for _, test := range tests {
+		have := test.level.String()
+		assert.Equal(t, test.want, have)
+	}
+}
+
+func TestFullName(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		names []string
+		want  string
+	}{
+		"empty": {},
+		"one name": {
+			names: []string{"a"},
+			want:  "a",
+		},
+		"multiple names": {
+			names: []string{"a", "b", "c"},
+			want:  "a.b.c",
+		},
+		"duplicate names": {
+			names: []string{"a", "a", "a"},
+			want:  "a.a.a",
+		},
+	}
+
+	for desc, test := range tests {
+		have := FullName(test.names)
+		assert.Equal(t, test.want, have, desc)
```

Comment:
Particularly in table driven testing: I was trained to log the function call / parameters that led to failure. E.g. here something like `FullName([]string{"a"}) = "b", want "a"` - the goal I believe was to enable as much context and debug as possible exclusively from logs without having to re-run the test to output context on the function called / its inputs?

Reply:
I just spotted that this advice is also on the [golang github repo](https://github.com/golang/go/issues/34384) FWIW (ctrl-f for `Identify the Function`).